content/augmentor.js

```diff
@@ -0,0 +1,198 @@
+// Copyright 2015, 2016 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+import { h, render } from 'preact';
+/** @jsx h */
+
+import { AugmentedIcon } from './components';
+
+import Runner from './runner';
+import { FETCH_IMAGE } from '../background/processor';
+
+import styles from './styles.less';
+
+export const AUGMENTED_NODE_ATTRIBUTE = 'data-parity-touched';
+
+export default class Augmentor {
+
+  static run (matches, resolved = {}) {
+    // Use the attributes matcher first
+    const attributesMatches = matches.filter((match) => match.from === 'attributes');
+    const textMatches = matches.filter((match) => match.from === 'text');
+
+    attributesMatches
+      .forEach((match) => {
+        const { email, name, node } = match;
+        const value = email || name;
+
+        Augmentor.augmentNode(value, node, resolved);
+      });
+
+    textMatches
+      .forEach((match) => {
+        const { email, name } = match;
+        const value = email || name;
+        const safeNode = Augmentor.getSafeNode(match);
+
+        Augmentor.augmentNode(value, safeNode, resolved);
+      });
+  }
+
+  static getSafeNode (match) {
+    const { email, node } = match;
+
+    // Safe Node is if the node which inner text is only the email address
+    if (node.innerText.trim() === email) {
+      return node;
+    }
+
+    const emailIndex = node.innerText.indexOf(email);
+
+    if (emailIndex === -1) {
+      return;
+    }
+
+    const beforeText = node.innerText.slice(0, emailIndex);
+    const afterText = node.innerText.slice(emailIndex + email.length);
+
+    const safeNode = document.createElement('span');
+    safeNode.innerText = email;
+
+    const nextNode = node.cloneNode(true);
+    nextNode.innerHTML = '';
+    nextNode.appendChild(safeNode);
+    safeNode.insertAdjacentText('beforebegin', beforeText);
+    safeNode.insertAdjacentText('afterend', afterText);
+
+    // Replace the node with the safe node
+    node.parentElement.replaceChild(nextNode, node);
+
+    return safeNode;
+  }
+
+  static augmentNode (key, node, resolved = {}) {
+    if (!node || node.getAttribute(AUGMENTED_NODE_ATTRIBUTE) === 'true') {
+      return;
+    }
+
+    const data = resolved[key];
+    node.setAttribute(AUGMENTED_NODE_ATTRIBUTE, true);
```

Comment:
But "no data" is cached for that node right? And when mutation observer kicks in it may change node content to something that would match (so we should probably trigger fetching again), but it's an edge case, don't think we should catter for that.